otter/convergence/logging.py

```diff
@@ -0,0 +1,141 @@
+"""Utilities for formatting log messages."""
+
+from collections import defaultdict
+from functools import partial
+
+from effect import parallel
+
+from pyrsistent import thaw
+
+from toolz.curried import groupby
+from toolz.itertoolz import concat
+
+from otter.convergence.steps import (
+    AddNodesToCLB, BulkAddToRCv3, BulkRemoveFromRCv3, ChangeCLBNode,
+    CreateServer, DeleteServer, RemoveNodesFromCLB, SetMetadataItemOnServer)
+from otter.log.cloudfeeds import cf_msg
+
+
+# - it kinda sucks that we're using separate effects for all of
+# these, maybe? CF has an API for sending a bunch of events at once and it'd be
+# good to use it. OTOH it could also be implemented at the logging observer
+# layer by using a "nagle".
+
+
+_loggers = {}
+
+
+def _logger(step_type):
+    """Declare a function for logging a bag of steps of a particular type."""
+    def _add_to_loggers(f):
+        _loggers[step_type] = f
+    return _add_to_loggers
+
+
+@_logger(CreateServer)
+def _(steps):
+    by_cfg = groupby(lambda s: s.server_config, steps)
+    effs = [
+        cf_msg(
+            'convergence-create-servers',
+            num_servers=len(cfg_steps),
+            server_config=dict(cfg))
+        # We sort the items with `thaw` because PMap does not support
+        # comparison
+        for cfg, cfg_steps in sorted(by_cfg.iteritems(), key=thaw)]
+    return parallel(effs)
+
+
+@_logger(SetMetadataItemOnServer)
```

Comment:
Non-blocking, because this is what we have so far, but I think we only use this to set DRAINING, and the fact that it's doing so by setting metadata is an implementation detail.  I wonder if we should bother with logging this at all right now, since it's not supported yet - and maybe we want to log something better?  But we can always just change it later, too.  :)